core-war/src/main/java/org/silverpeas/web/directory/control/DirectorySessionController.java

```diff
@@ -799,4 +810,91 @@ private DirectoryItem getDirectoryItem(final DirectoryItemList allUsers,
     }
     return allUsers.getItemByUniqueId(itemId);
   }
+
+  public Form getExtraForm() {
+    PublicationTemplate template = getExtraTemplate();
+    if (template != null) {
+      try {
+        Form searchForm = template.getSearchForm();
+        if (xmlData != null) {
+          searchForm.setData(xmlData);
+        } else {
+          searchForm.setData(template.getRecordSet().getEmptyRecord());
+        }
+        return searchForm;
+      } catch (Exception e) {
+        SilverLogger.getLogger(this).error(e);
+      }
+    }
+    return null;
+  }
+
+  private PublicationTemplate getExtraTemplate() {
+    if (xmlTemplate == null && !xmlTemplateLoaded) {
+      PublicationTemplateManager templateManager = PublicationTemplateManager.getInstance();
+      xmlTemplate = templateManager.getDirectoryTemplate();
+      xmlTemplateLoaded = true;
+    }
+    return xmlTemplate;
+  }
+
+  private void saveExtraRequest(List<FileItem> items) {
+    PublicationTemplateImpl template = (PublicationTemplateImpl) getExtraTemplate();
+    if (template != null) {
+      // build a dataRecord object storing user's entries
+      try {
+        RecordTemplate searchTemplate = template.getSearchTemplate();
+        DataRecord data = searchTemplate.getEmptyRecord();
+
+        PagesContext context = new PagesContext("useless", "useless", getLanguage(), getUserId());
+
+        XmlSearchForm searchForm = (XmlSearchForm) template.getSearchForm();
+        searchForm.update(items, data, context);
+
+        // xmlQuery is in the data object, store it into session
+        xmlData = data;
+      } catch (Exception e) {
+        SilverLogger.getLogger(this).error(e);
+      }
+    }
+  }
+
+  public QueryDescription buildQuery(List<FileItem> items) {
+    String query = FileUploadUtil.getParameter(items, "key");
+    QueryDescription queryDescription = new QueryDescription(query);
+    queryDescription.addComponent("users");
+    for (String appId : getContactComponentIds()) {
+      queryDescription.addComponent(appId);
+    }
+
+    setCurrentQuery(query);
+    saveExtraRequest(items);
+
+    buildExtraQuery(queryDescription);
+    return queryDescription;
+  }
+
+  private void buildExtraQuery(QueryDescription query) {
+    PublicationTemplateImpl template = (PublicationTemplateImpl) getExtraTemplate();
+    if (template != null) {
+      // build the xmlSubQuery according to the dataRecord object
+      String templateFileName = template.getFileName();
+      String templateName = templateFileName.substring(0, templateFileName.lastIndexOf("."));
+      if (xmlData != null) {
+        for (String fieldName : xmlData.getFieldNames()) {
+          try {
+            Field field = xmlData.getField(fieldName);
+            String fieldValue = field.getStringValue();
+            if (fieldValue != null && fieldValue.trim().length() > 0) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Refactor this code to not nest more than 4 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS134)